exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/xsort/managed/ExternalSortBatch.java

```diff
@@ -0,0 +1,1321 @@
+
+    CopierHolder.BatchMerger merger = copierHolder.startFinalMerge(schema, allBatches, container, outputBatchRecordCount);
+    merger.next();
+    resultsIterator = merger;
+    sortState = SortState.DELIVER;
+    return IterOutcome.OK_NEW_SCHEMA;
+  }
+
+  private boolean consolidateBatches() {
+
+    // Determine additional memory needed to hold one batch from each
+    // spilled run.
+
+    int inMemCount = bufferedBatches.size();
+    int spilledRunsCount = spilledRuns.size();
+
+    // Can't merge more than will fit into memory at one time.
+
+    int maxMergeWidth = (int) (mergeMemoryPool / estimatedOutputBatchSize);
+    maxMergeWidth = Math.min(mergeLimit, maxMergeWidth);
+
+    // If we can't fit all batches in memory, must spill any in-memory
+    // batches to make room for multiple spill-merge-spill cycles.
+
+    if (inMemCount > 0) {
+      if (spilledRunsCount > maxMergeWidth) {
+        spillFromMemory();
+        return true;
+      }
+
+      // If we just plain have too many batches to merge, spill some
+      // in-memory batches to reduce the burden.
+
+      if (inMemCount + spilledRunsCount > mergeLimit) {
+        spillFromMemory();
+        return true;
+      }
+
+      // If the on-disk batches and in-memory batches need more memory than
+      // is available, spill some in-memory batches.
+
+      long allocated = allocator.getAllocatedMemory();
+      long totalNeeds = spilledRunsCount * estimatedOutputBatchSize + allocated;
+      if (totalNeeds > mergeMemoryPool) {
+        spillFromMemory();
+        return true;
+      }
+    }
+
+    // Merge on-disk batches if we have too many.
+
+    int mergeCount = spilledRunsCount - maxMergeWidth;
+    if (mergeCount <= 0) {
+      return false;
+    }
+
+    // We will merge. This will create yet another spilled
+    // run. Account for that.
+
+    mergeCount += 1;
+
+    // Must merge at least 2 batches to make progress.
+
+    mergeCount = Math.max(2, mergeCount);
+
+    mergeCount = Math.min(mergeCount, maxMergeWidth);
+
+    // Do the merge, then loop to try again in case not
+    // all the target batches spilled in one go.
+
+    logger.trace("Merging {} on-disk runs, Alloc. memory = {}",
+        mergeCount, allocator.getAllocatedMemory());
+    mergeAndSpill(spilledRuns, mergeCount);
+    return true;
+  }
+
+  /**
+   * This operator has accumulated a set of sorted incoming record batches.
+   * We wish to spill some of them to disk. To do this, a "copier"
+   * merges the target batches to produce a stream of new (merged) batches
+   * which are then written to disk.
+   * <p>
+   * This method spills only half the accumulated batches
+   * minimizing unnecessary disk writes. The exact count must lie between
+   * the minimum and maximum spill counts.
+    */
+
+  private void spillFromMemory() {
+    int spillCount;
+    if (spillFileSize == 0) {
+      // If no spill limit given, guess half the batches.
+
+      spillCount = bufferedBatches.size() / 2;
```

Comment:
Actually, this line is here only as a fall back. Notice the surrounding if. We are, in fact, driven by the spill file size.

Changed this to set a minimum spill file size in setup, then eliminate this line here.